README.md
# iuvui - intel-undervolt user interface

GUI for [intel-undervolt](https://github.com/kitsunyan/intel-undervolt)

## Screenshot
<img src="resources/iuvui_screenshot.png" width=554>

## Dependencies & Requirements
- intel-undervolt
- root

## GUI based on:
- [nuklear](https://github.com/vurtun/nuklear)

## Running:
- make & run as root

## TODO:
- add histogram (or similar) for power consumption


CHANGELOG.md
# v0.3.0
    - added: info about current temperatures

# v0.2.0
    - added: info about current power consumption

# v0.1.0
    - added: enable/disable on boot


src/measure.c
#include "measure.h"

void savePowercapNextAsDouble(powercap_list_t *lst, int maxname) {
	char buf[BUFSZSMALL];
	if (!lst) {
		return;
	}

	savePowercapNextAsDouble(lst->next, maxname);

	sprintf(buf, DIR_POWERCAP "/%s/energy_uj", lst->dir);
	int fd = open(buf, O_RDONLY);
	if (fd >= 0) {
		int count = read(fd, buf, BUFSZSMALL - 1);
		if (count > 0) {
			buf[buf[count - 1] == '\n' ? count - 1 : count] = '\0';
			int64_t val = (int64_t) atoll(buf);
			struct timespec tnow;
			clock_gettime(CLOCK_MONOTONIC, &tnow);
			if (lst->last > 0) {
				struct timespec tdiff;
				tdiff.tv_sec = tnow.tv_sec - lst->time.tv_sec;
				tdiff.tv_nsec = tnow.tv_nsec - lst->time.tv_nsec;
				while (tdiff.tv_nsec < 0) {
					tdiff.tv_nsec += 1000000000;
					tdiff.tv_sec--;
				}
				int64_t diff = (int64_t) (tdiff.tv_sec * 1000000000 + tdiff.tv_nsec);
				lst->val = (double) (val - lst->last) * 1000 / diff;
			}
			lst->last = val;
			lst->time = tnow;
		}
		close(fd);
	}
}

powercap_list_t *getPowercap(int *maxname) {
	char buf[BUFSZSMALL];
	powercap_list_t * lst = NULL;

	DIR * dir = opendir(DIR_POWERCAP);
	if (dir == NULL) {
		fprintf(stderr, "Failed to open powercap directory\n");
		return NULL;
	}
	struct dirent * dirent;
	while ((dirent = readdir(dir))) {
		if (strstr(dirent->d_name, ":") && strlen(dirent->d_name) <= 30) {
			sprintf(buf, DIR_POWERCAP "/%s/name", dirent->d_name);
			int fd = open(buf, O_RDONLY);
			if (fd >= 0) {
				int count = read(fd, buf, BUFSZSMALL - 1);
				if (count > 1) {
					int nlen = buf[count - 1] == '\n' ? count - 2 : count - 1;
					buf[nlen + 1] = '\0';
					powercap_list_t * nlst = malloc(sizeof(powercap_list_t));
					nlst->next = lst;
					lst = nlst;
					nlst->name = malloc(nlen + 1);
					memcpy(nlst->name, buf, nlen + 1);
					int dlen = strlen(dirent->d_name);
					nlst->dir = malloc(dlen + 1);
					memcpy(nlst->dir, dirent->d_name, dlen + 1);
					nlst->last = 0;
					if (maxname) {
						*maxname = nlen > *maxname ? nlen : *maxname;
					}
				}
				close(fd);
			}
		}
	}
	closedir(dir);

	return lst;
}

void getHwmonNextValues(hwmon_list_t *lst, int maxname, char *buf) {
	if (!lst) {
		return;
	}

	getHwmonNextValues(lst->next, maxname, buf);

	sprintf(buf, DIR_HWMON "/%s/temp%d_input", lst->dir, lst->index);
	int fd = open(buf, O_RDONLY);
	if (fd >= 0) {
		int count = read(fd, buf, BUFSZSMALL - 1);
		if (count > 0) {
			buf[buf[count - 1] == '\n' ? count - 1 : count] = '\0';
			double dval = atol(buf) / 1000.;
			snprintf(lst->val, BUFSZSMALL, "%6.01f%s", dval, " C");
		}
		close(fd);
	}
}

bool getHwmon(const char * name, char * out) {
	char buf[BUFSZSMALL];

	DIR * dir = opendir(DIR_HWMON);
	if (dir == NULL) {
		fprintf(stderr, "Failed to open hwmon directory\n");
		return NULL;
	}
	struct dirent * dirent;
	while ((dirent = readdir(dir))) {
		if (strlen(dirent->d_name) <= 30) {
			sprintf(buf, DIR_HWMON "/%s/name", dirent->d_name);
			int fd = open(buf, O_RDONLY);
			if (fd >= 0) {
				int count = read(fd, buf, BUFSZSMALL - 1);
				if (count > 1) {
					int nlen = buf[count - 1] == '\n' ? count - 2 : count - 1;
					buf[nlen + 1] = '\0';
					if (!strcmp(buf, name)) {
						strcpy(out, dirent->d_name);
						close(fd);
						closedir(dir);
						return true;
					}
				}
				close(fd);
			}
		}
	}
	closedir(dir);

	return false;
}

hwmon_list_t *getCoretemp(int *maxname) {
	char hdir[BUFSZSMALL];
	char buf[BUFSZSMALL];
	hwmon_list_t * lst = NULL;

	if (!getHwmon("coretemp", hdir)) {
		fprintf(stderr, "Failed to find coretemp hwmon\n");
		return NULL;
	}

	int i;
	for (i = 1;; i++) {
		sprintf(buf, DIR_HWMON "/%s/temp%d_input", hdir, i);
		int fd = open(buf, O_RDONLY);
		if (fd < 0) {
			break;
		}
		sprintf(buf, DIR_HWMON "/%s/temp%d_label", hdir, i);
		fd = open(buf, O_RDONLY);
		char * name = NULL;
		if (fd >= 0) {
			int count = read(fd, buf, BUFSZSMALL - 1);
			if (count > 1) {
				int nlen = buf[count - 1] == '\n' ? count - 2 : count - 1;
				buf[nlen + 1] = '\0';
				name = malloc(strlen(buf) + 1);
				strcpy(name, buf);
			}
			close(fd);
		}
		if (!name) {
			sprintf(buf, "temp%d", i);
			name = malloc(strlen(buf) + 1);
			strcpy(name, buf);
		}
		hwmon_list_t * nlst = malloc(sizeof(hwmon_list_t));
		nlst->next = lst;
		lst = nlst;
		nlst->name = name;
		strncat(nlst->name, " : ", 4);
		nlst->dir = malloc(strlen(hdir) + 1);
		strcpy(nlst->dir, hdir);
		nlst->index = i;
		if (maxname) {
			int len = strlen(name);
			*maxname = len > *maxname ? len : *maxname;
		}
	}

	return lst;
}

/*
void printCpufreq(int maxname, char *buf, bool *nl) {
	bool nll = false;

	int i;
	for (i = 0;; i++) {
		sprintf(buf, "/sys/bus/cpu/devices/cpu%d/cpufreq/scaling_cur_freq", i);
		int fd = open(buf, O_RDONLY);
		if (fd < 0) {
			break;
		}
		int count = read(fd, buf, BUFSZSMALL - 1);
		if (count > 0) {
			buf[buf[count - 1] == '\n' ? count - 1 : count] = '\0';
			double dval = atol(buf) / 1000.;
			sprintf(buf, "Core %d", i);
			printf("%9.03f MHz\n", dval);
		}
		close(fd);
	}
}
*/


src/measure.h
/* taken from https://github.com/kitsunyan/intel-undervolt
 * and made a few modifications
 */

#include <dirent.h>
#include <fcntl.h>
#include <iconv.h>
#include <langinfo.h>
#include <locale.h>
#include <signal.h>
#include <stdbool.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>

#define DIR_POWERCAP "/sys/class/powercap"
#define DIR_HWMON "/sys/class/hwmon"
#define BUFSZSMALL 80

typedef struct {
	void * next;
	char * name;
	char * dir;
	int64_t last;
    double val;
	struct timespec time;
} powercap_list_t;

typedef struct {
	void * next;
	char * name;
	char * dir;
	char val[BUFSZSMALL];
	int index;
} hwmon_list_t;

void savePowercapNextAsDouble(powercap_list_t *, int);
void printCpufreq(int, char *, bool *);
void getHwmonNextValues(hwmon_list_t *, int, char *);
bool getHwmon(const char *, char *);
powercap_list_t *getPowercap(int *);
hwmon_list_t *getCoretemp(int *);


src/intel_undervolt.h
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>

#include "measure.h"

#define CONFFILE        "/etc/intel-undervolt.conf"

/* count elements below */
#define COUNTELEMENTS   6

#define CPUOFFSET       0
#define GPUOFFSET       1
#define CPUCACHEOFFSET  2
#define SYSAOFFSET      3
#define ANALOGIOOFFSET  4
#define DAEMONINTERVAL  5

/* max number of digits */
#define MAXDIGIT        6
/* Length for undervolt option */
#define UV_OPTION       64

#define MAXLENSENTENCE  128

/* max and minimal voltage offset */
#define MAXVOLTAGEOFF   100
#define MINVOLTAGEOFF   -250

#define BUFFERSIZE      4096

#define VOLTAGESTEP     1

/*
 * Set the value into the config file /etc/intel-undervolt.conf
 * and apply it. Returns 0 on success
 */
int setValIntoConfFile(int, short);

/*
 * Returns 0 on success, -1 on error
 * TODO: add support for the following settings:
 *    power package ..
 *    tjoffset
 *    interval
 */
int readVal(int *);

/* 
 * applyValues - set values into conf file and apply.
 * Also compares with current values in the conf file
 * /etc/intel-undervolt.conf , if the value is the same,
 * no change will be made to that one.
 * Returns how many values have been changed.
 */
short applyValues(int *, int *);

int measurePowerConsumption(char [][BUFSZSMALL], powercap_list_t *, int, int);

int getCpuTemp(char [][BUFSZSMALL], hwmon_list_t *, int *, int);

/*
 * Set Power Limits Alteration
 * see config:
 *     # Usage: power ${domain} ${short_power_value} ${long_power_value}
 *     # Power value: ${power}[/${time_window}][:enabled][:disabled]
 *     # Supported domains: package
 *        power package 45 35
 *        power package 45/0.002 35/28
 *        power package 45/0.002:disabled 35/28:enabled
 * Returns 0 on success, -1 on error
 */
int powerLimitAlt(float, float);

/*
 * Set Critical Temperature Offset Alteration
 * see config:
 *    # Usage: tjoffset ${temperature_offset}
 *        tjoffset -20
 * Returns 0 on success, -1 on error
 */
int tempOffsetAlt(int);

/*
 * Set Daemon Update Interval
 * see config:
 *    # Usage: interval ${interval_in_milliseconds}
 *        interval 5000
 * Returns 0 on success, -1 on error
 */
int daemonUpdateInterval(unsigned int);

/*
 * Enable/disable daemon.
 * int set = 0 -> disable,
 * int set = 1 -> enable,
 * int set = 2 -> check if enabled,
 * returns 0 on success, -1 on error
 */
int systemdService(int);

/*
 * Resets all Values to 0
 * Returns 0 on success, -1 on error
 */
int uvResetAll();

/*
 * convert int to string
 */
void intToString(int , char *, char *);


src/intel_undervolt.c
#include "intel_undervolt.h"

int setValIntoConfFile(int offset, short option) {
        char newConf[MAXLENSENTENCE];
        char tmp[MAXLENSENTENCE];
        char currConf[MAXLENSENTENCE];
        char *fileSize;
        struct stat buf; 

        fpos_t fpos;

        FILE *f = fopen(CONFFILE, "r+");

        if (!f) {
                return 1;
        }

        switch (option)
        {
                case CPUOFFSET:
                        strncpy(newConf, "undervolt 0 'CPU' ", 19);
                        intToString(offset, tmp, newConf);
                        break;
                case GPUOFFSET:
                        strncpy(newConf, "undervolt 1 'GPU' ", 19);
                        intToString(offset, tmp, newConf);
                        break;
                case CPUCACHEOFFSET:
                        strncpy(newConf, "undervolt 2 'CPU Cache' ", 25);
                        intToString(offset, tmp, newConf);
                        break;
                case SYSAOFFSET:
                        strncpy(newConf, "undervolt 3 'System Agent' ", 28);
                        intToString(offset, tmp, newConf);
                        break;
                case ANALOGIOOFFSET:
                        strncpy(newConf, "undervolt 4 'Analog I/O' ", 26);
                        intToString(offset, tmp, newConf);
                        break;
                case DAEMONINTERVAL:
                        strncpy(newConf, "interval ", 10);
                        intToString(offset, tmp, newConf);
                        break;
                default:
                        return 1;
        }
        strncat(newConf, "\n", 2);

        while (!feof(f)) {
                fgetpos(f, &fpos);
                fgets(currConf, MAXLENSENTENCE, f);

                if (strchr(currConf, '#') || strlen(currConf) < 8) {
                        continue;
                } else if (option == CPUOFFSET && strstr(currConf, "undervolt 0 \'CPU")) {
                        break;
                } else if (option == GPUOFFSET && strstr(currConf, "undervolt 1 \'GPU")) {
                        break;
                } else if (option == CPUCACHEOFFSET && strstr(currConf, "undervolt 2 \'CPU Cache")) {
                        break;
                } else if (option == SYSAOFFSET && strstr(currConf, "undervolt 3 \'System Agent")) {
                        break;
                } else if (option == ANALOGIOOFFSET && strstr(currConf, "undervolt 4 \'Analog I/O")) {
                        break;
                } else if (option == DAEMONINTERVAL && strstr(currConf, "interval")) {
                        break;
                }
        }

        /* get file size and allocate memory for it */
        fstat(fileno(f), &buf);
        fileSize = (char *)calloc((size_t)buf.st_size, sizeof(char));
        fread(fileSize, sizeof(char), (size_t)buf.st_size, f);

        fsetpos(f, &fpos);
        fputs(newConf, f);
        fputs(fileSize, f);

        fclose(f);
        free(fileSize);
        strcpy(newConf, "");

        return 0;
}

int readVal(int *currVal) {
        char readCfg[MAXLENSENTENCE];

        FILE *f = fopen(CONFFILE, "r");

        if (!f) {
                return 1;
        }
        while (!feof(f)) {
                fgets(readCfg, 128, f);

                if (strchr(readCfg, '#')) {
                        continue;
                } else if (strstr(readCfg, "undervolt 0 \'CPU\'")) {
                        currVal[CPUOFFSET] = (int)strtof(strrchr(readCfg, ' ')+1, NULL);
                } else if (strstr(readCfg, "undervolt 1 \'GPU\'")) {
                        currVal[GPUOFFSET] = (int)strtof(strrchr(readCfg, ' ')+1, NULL);
                } else if (strstr(readCfg, "undervolt 2 \'CPU Cache\' ")) {
                        currVal[CPUCACHEOFFSET] = (int)strtof(strrchr(readCfg, ' ')+1, NULL);
                } else if (strstr(readCfg, "undervolt 3 \'System Agent\'")) {
                        currVal[SYSAOFFSET] = (int)strtof(strrchr(readCfg, ' ')+1, NULL);
                } else if (strstr(readCfg, "undervolt 4 \'Analog I/O\'")) {
                        currVal[ANALOGIOOFFSET] = (int)strtof(strrchr(readCfg, ' ')+1, NULL);
                } else if (strstr(readCfg, "interval")) {
                        currVal[DAEMONINTERVAL] = (int)strtof(strrchr(readCfg, ' ')+1, NULL);
                }
        }

        return 0;
}

short applyValues(int *newValues, int *currentValues) {
        FILE *fproc;
        /* count how much values have changed
         * if zero, non has changed and nothing has been set
         * if -1, error while applying config
         */
        short count = 0;

        if (newValues[CPUOFFSET] != currentValues[CPUOFFSET]) {
                if (setValIntoConfFile(newValues[CPUOFFSET], CPUOFFSET))
                        return -1;
                count++;
                currentValues[CPUOFFSET] = newValues[CPUOFFSET];
        }
        if (newValues[GPUOFFSET] != currentValues[GPUOFFSET]) {
                if (setValIntoConfFile(newValues[GPUOFFSET], GPUOFFSET))
                        return -1;
                count++;
                currentValues[GPUOFFSET] = newValues[GPUOFFSET];
        }
        if (newValues[CPUCACHEOFFSET] != currentValues[CPUCACHEOFFSET]) {
                if (setValIntoConfFile(newValues[CPUCACHEOFFSET], CPUCACHEOFFSET))
                        return -1;
                count++;
                currentValues[CPUCACHEOFFSET] = newValues[CPUCACHEOFFSET];
        }
        if (newValues[SYSAOFFSET] != currentValues[SYSAOFFSET]) {
                if (setValIntoConfFile(newValues[SYSAOFFSET], SYSAOFFSET))
                        return -1;
                count++;
                currentValues[SYSAOFFSET] = newValues[SYSAOFFSET];
        }
        if (newValues[ANALOGIOOFFSET] != currentValues[ANALOGIOOFFSET]) {
                if (setValIntoConfFile(newValues[ANALOGIOOFFSET], ANALOGIOOFFSET))
                        return -1;
                count++;
                currentValues[ANALOGIOOFFSET] = newValues[ANALOGIOOFFSET];
        }
        if (newValues[DAEMONINTERVAL] != currentValues[DAEMONINTERVAL]) {
                if (setValIntoConfFile(newValues[DAEMONINTERVAL], DAEMONINTERVAL))
                        return -1;
                count++;
                currentValues[DAEMONINTERVAL] = newValues[DAEMONINTERVAL];
        }

        /* write to file and apply settings */
        if ((fproc = popen("intel-undervolt apply", "r")) == NULL) {
                count = -1;
        }

        pclose(fproc);
        return count;
}

int measurePowerConsumption(char currPVals[][BUFSZSMALL], powercap_list_t *pcapList, int elCount, int maxname) {
        int i = 0;
        powercap_list_t *pListNext = pcapList;

        savePowercapNextAsDouble(pcapList, maxname);

        while (pListNext) {
                if (i < elCount) {
                        snprintf(currPVals[i], BUFSZSMALL, "%s : %3.03f W", pListNext->name, pListNext->val);
                        ++i;
                } else {
                        break;
                }
                pListNext = pListNext->next;
        }

        return 0;
}

int getCpuTemp(char currTempVals[][BUFSZSMALL], hwmon_list_t *hwlst, int *maxname, int el){
        char tmp[BUFSZSMALL];
        int i = 0;
        hwlst = getCoretemp(&maxname);
        if (hwlst != NULL) {
                getHwmonNextValues(hwlst, maxname, tmp);
                hwmon_list_t *hwmonNext = hwlst;
                while (hwmonNext && i < el) {
                        snprintf(currTempVals[i], BUFSZSMALL, "%s %s", hwmonNext->name, hwmonNext->val);
                        hwmonNext = hwmonNext->next;
                        ++i;
                }
        }
        return 0;
}

int powerLimitAlt(float shortPowerVal, float longPowerVal) {
        /* TODO */
        return 0;
}

int tempOffsetAlt(int temp) {
        /* TODO */
        return 0;
}

int daemonUpdateInterval(unsigned int time) {
        return setValIntoConfFile(time, DAEMONINTERVAL);
}

int systemdService(int set) {
        FILE *f;
        char tmpTxt[BUFFERSIZE];

        if (set) {
                if ((f = popen("sudo systemctl enable intel-undervolt", "r")) == NULL) {
                        return -1;
                }
                pclose(f);
                if ((f = popen("sudo systemctl start intel-undervolt", "r")) == NULL) {
                        return -1;
                }
        } else if (set == 0) {
                if ((f = popen("sudo systemctl disable intel-undervolt", "r")) == NULL) {
                        return -1;
                }
                pclose(f);
                if ((f = popen("sudo systemctl stop intel-undervolt", "r")) == NULL) {
                        return -1;
                }
        } else {
                if ((f = popen("sudo systemctl status intel-undervolt", "r")) == NULL) {
                        return -1;
                }
                while (fgets(tmpTxt, BUFFERSIZE, f) != NULL) {
                        if (strstr(tmpTxt, "intel-undervolt.service; enabled") != NULL) {
                                pclose(f);
                                return 0;
                        }
                }
        }
        pclose(f);

        if ((f = popen("sudo systemctl status intel-undervolt", "r")) == NULL) {
                return -1;
        }
        while (fgets(tmpTxt, BUFFERSIZE, f) != NULL) {
                if (strstr(tmpTxt, "status=0/SUCCESS")) {
                        pclose(f);
                        return 0;
                } else if (strstr(tmpTxt, "intel-undervolt.service; disabled") != NULL) {
                        pclose(f);
                        return 0;
                }
        }
        pclose(f);
        return -1;
}

int uvResetAll() {
        char *resetValues = "# CPU Undervolting\n"
                        "# Usage: undervolt ${index} ${display_name} ${undervolt_value}\n"
                        "# Example: undervolt 2 'CPU Cache' -25.84\n\n"
                        "undervolt 0 'CPU' 0\n"
                        "undervolt 1 'GPU' 0\n"
                        "undervolt 2 'CPU Cache' 0\n"
                        "undervolt 3 'System Agent' 0\n"
                        "undervolt 4 'Analog I/O' 0\n\n"
                        "# Power Limits Alteration\n"
                        "# Usage: power ${domain} ${short_power_value} ${long_power_value}\n"
                        "# Power value: ${power}[/${time_window}][:enabled][:disabled]\n"
                        "# Supported domains: package\n"
                        "# Example: power package 45 35"
                        "# Example: power package 45/0.002 35/28\n"
                        "# Example: power package 45/0.002:disabled 35/28:enabled\n\n"
                        "# Critical Temperature Offset Alteration\n"
                        "# Usage: tjoffset ${temperature_offset}\n"
                        "# Example: tjoffset -20\n\n"
                        "# Daemon Update Interval\n"
                        "# Usage: interval ${interval_in_milliseconds}\n\n"
                        "interval 5000\n";
        FILE *f = fopen(CONFFILE, "w");

        if (!f) {
                return -1;
        }
        if (fprintf(f, "%s", resetValues) < 0) {
                fclose(f);
                return -1;
        }
        fclose(f);

        /* write to file and apply settings */
        if((f = popen("intel-undervolt apply", "r")) == NULL) {
                return -1;
        }

        pclose(f);

        return 0;
}

void intToString(int offset, char *tmp, char *dest) {
        snprintf(tmp, MAXDIGIT, "%d", offset);
        strncat(dest, tmp, UV_OPTION);
}
